src/tests/system/tests/test_access_control_ldap_filter.py

```diff
@@ -0,0 +1,254 @@
+        2. Create `user2` without an email attribute.
+        3. Configure SSSD with `access_provider = ldap|ad` and `*_access_filter = mail=*domain.com`.
+        4. Start SSSD.
+    :steps:
+        1. Attempt login with `user1` (matching the wildcard filter).
+        2. Attempt login with `user2` (lacking the attribute).
+    :expectedresults:
+        1. `user1` is successfully authenticated.
+        2. `user2` is denied access.
+    :customerscenario: False
+    """
+
+    # Revisit this logic once https://github.com/SSSD/sssd-test-framework/pull/161 is merged
+    if isinstance(provider, LDAP):
+        provider.user("user1").add(mail="user1@domain.com")
+    else:
+        provider.user("user1").add(email="user1@domain.com")
+
+    provider.user("user2").add()
+
+    # None of the filters below work :/
+    if isinstance(provider, AD) or isinstance(provider, Samba):
+        client.sssd.domain["access_provider"] = "ad"
+        client.sssd.domain["ad_access_filter"] = "(mail=*domain.com)"
+    elif isinstance(provider, IPA):
+        client.sssd.domain["access_provider"] = "ldap"
+        client.sssd.domain["ldap_access_filter"] = "(mail=*domain.com)"
+    elif isinstance(provider, LDAP):
+        client.sssd.domain["access_provider"] = "ldap"
+        client.sssd.domain["ldap_access_filter"] = "(mail=*domain.com)"
+    else:
+        raise RuntimeError("Provider not supported")
+
+    client.sssd.start()
+
+    assert client.auth.ssh.password("user1", "Secret123"), "User1 should be able to log in!"
+    assert not client.auth.ssh.password("user2", "Secret123"), "User2 should NOT be able to log in!"
+
+
+@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+@pytest.mark.importance("critical")
+def test_access_filter__ldap_query_with_and_or_not_permits_user_login(client: Client, provider: GenericProvider):
+    """
+    :title: LDAP query with AND, OR, NOT operators permits user login
+    :description: Verifies that SSSD correctly processes LDAP queries using `AND (&)`, `OR (|)`, and `NOT (!)`.
+    :setup:
+        1. Create user `user1` (email: user1@domain.com, uid: 10001)
+        2. Create user `user2` (email: user2@domain.com, uid: 10002)
+        3. Create user `user3` (email: user3@example.com, uid: 10003)
+        4. Configure SSSD with `access_provider = ldap|ad` and `*_access_filter` that:
+            - Allows users with `@domain.com` email OR uid `10001`
+            - Denies users with uid `10003`
+        5. Start SSSD
+    :steps:
+        1. Attempt login with `user1`
+        2. Attempt login with `user2`
+        3. Attempt login with `user3`
+    :expectedresults:
+        1. `user1` is successfully authenticated.
+        2. `user2` is denied access.
+        3. `user3` is denied access.
+    :customerscenario: False
+    """
+
+    # Revisit this logic once https://github.com/SSSD/sssd-test-framework/pull/161 is merged
+    if isinstance(provider, LDAP):
```

Comment:
for this, something like this,
```
email_key = "mail" if isinstance(provider, LDAP) else "email"

users = [
    ("user1", "user1@domain.com", 10001),
    ("user2", "user2@domain.com", 10002),
    ("user3", "user3@example.com", 10003),
]

for username, email, uid in users:
    provider.user(username).add(**{email_key: email, "uid": uid})

```

Reply:
I kept it smaller, wdyt?

```
    email_key = "mail" if isinstance(provider, LDAP) else "email"
    provider.user("user1").add(**{email_key: "user1@domain.com"}, uid=10001)
    provider.user("user2").add(**{email_key: "user2@domain.com"}, uid=10002)
    provider.user("user3").add(**{email_key: "user3@example.com"}, uid=10003)
```

Reply:
Uses a list of tuples to store user details, avoiding repetitive calls.
Loops through users to eliminate redundant lines.
Keeps the logic clean and scalable.
So, I will go with for loop.

Reply:
I decided to not use the dictionary to avoid mypy issues.

Reply:
Yes, I did not find any good solution for this.